chain/network/src/peer_manager/network_state/tier1.rs

```diff
@@ -77,4 +179,147 @@ impl super::NetworkState {
         self.config.event_sink.push(Event::Tier1AdvertiseProxies(new_data.clone()));
         new_data
     }
+
+    /// Closes TIER1 connections from nodes which are not TIER1 any more.
+    /// If this node is TIER1, it additionally connects to proxies of other TIER1 nodes.
+    pub async fn tier1_connect(self: &Arc<Self>, clock: &time::Clock) {
+        let tier1_cfg = match &self.config.tier1 {
+            Some(it) => it,
+            None => return,
+        };
+        if !tier1_cfg.enable_outbound {
+            return;
+        }
+        let accounts_data = self.accounts_data.load();
+        let validator_cfg = self.tier1_validator_config(&accounts_data);
+
+        // Construct indices on accounts_data.
+        let mut accounts_by_proxy = HashMap::<_, Vec<_>>::new();
+        let mut proxies_by_account = HashMap::<_, Vec<_>>::new();
+        for d in accounts_data.data.values() {
+            proxies_by_account.entry(&d.account_key).or_default().extend(d.proxies.iter());
+            for p in &d.proxies {
+                accounts_by_proxy.entry(&p.peer_id).or_default().push(&d.account_key);
+            }
+        }
+
+        // Browse the connections from newest to oldest.
+        let tier1 = self.tier1.load();
+        let mut ready: Vec<_> = tier1.ready.values().collect();
+        ready.sort_unstable_by_key(|c| c.established_time);
+        ready.reverse();
+
+        // Select the oldest TIER1 connection for each account.
+        let mut safe = HashMap::<&PublicKey, &PeerId>::new();
+        if validator_cfg.is_some() {
+            // TIER1 nodes can also connect to TIER1 proxies.
+            for conn in &ready {
+                let peer_id = &conn.peer_info.id;
+                for key in accounts_by_proxy.get(peer_id).into_iter().flatten() {
+                    safe.insert(key, peer_id);
+                }
+            }
+        }
+        // Direct TIER1 connections have priority.
+        for key in &accounts_data.keys {
+            if let Some(conn) = tier1.ready_by_account_key.get(&key) {
+                safe.insert(key, &conn.peer_info.id);
```

Comment:
what should happen if validator_cfg  is None?

should we still add these connections to the safe set? 

I see that you do multiple times 'if Some(vc) = validator_cfg -- shouldn't we do this verification at the beginning of the method (and AFAIK - if validator_cfg is None- simply close all the connections and be done?)

Reply:
no, TIER1 proxies are not TIER1 nodes, but they also accept TIER1 connections.